src/main/java/ee/carlrobert/codegpt/codecompletions/InlineCodeCompletionRenderer.java

```diff
@@ -0,0 +1,226 @@
+import ee.carlrobert.codegpt.conversations.ConversationService;
+import ee.carlrobert.codegpt.conversations.message.Message;
+import ee.carlrobert.codegpt.settings.configuration.ConfigurationState;
+import ee.carlrobert.codegpt.settings.service.FillInTheMiddle;
+import ee.carlrobert.codegpt.settings.state.SettingsState;
+import java.awt.Font;
+import java.awt.FontMetrics;
+import java.awt.Graphics2D;
+import java.awt.event.KeyEvent;
+import java.awt.event.KeyListener;
+import java.awt.geom.Rectangle2D;
+import javax.swing.Timer;
+import org.jetbrains.annotations.NotNull;
+
+public class InlineCodeCompletionRenderer implements EditorCustomElementRenderer, DocumentListener,
+    KeyListener, CaretListener, EditorMouseListener, SelectionListener {
+
+  public static final Key<Inlay<InlineCodeCompletionRenderer>> INLAY_KEY = Key.create(
+      "codegpt.inlay");
+  public static final String ACTION_ID = "InsertInlineTextAction";
+  public static final int MAX_OFFSET = 400; // TODO: Add in settings? Always entire file?
+
+  private Timer typingTimer;
+  private String inlayText = "";
+  private CompletionRequestHandler requestHandler;
+
+  private final Editor editor;
+  private final ConversationService conversationService;
+  private final Conversation conversation;
+
+
+  public InlineCodeCompletionRenderer(Editor editor) {
+    this.editor = editor;
+    ((EditorEx) editor).getDocument().addDocumentListener(this);
+    editor.getContentComponent().addKeyListener(this);
+    editor.addEditorMouseListener(this);
+    editor.getCaretModel().addCaretListener(this);
+    editor.getSelectionModel().addSelectionListener(this);
+    initTypingTimer();
+    this.conversationService = ConversationService.getInstance();
+    var completionCode =
+        SettingsState.getInstance().getSelectedService().getCompletionCode() + ".inline";
+    this.conversation = conversationService.createConversation(completionCode);
+    conversationService.addConversation(this.conversation);
+  }
+
+  private void initTypingTimer() {
+    typingTimer = new Timer(ConfigurationState.getInstance().getInlineDelay(),
+        e -> ApplicationManager.getApplication().invokeLater(() -> {
+          if (editor != null && !editor.isDisposed()) {
+            Inlay<InlineCodeCompletionRenderer> inlay = editor.getUserData(INLAY_KEY);
+            if (inlay == null) {
+              Document document = editor.getDocument();
+              if (document.getText().isEmpty() || document.getText().isBlank()) {
+                return;
+              }
+              int offset = editor.getCaretModel().getOffset();
+              var message = createFimMessage(offset, document);
+              typingTimer.stop();
+              requestHandler = new CompletionRequestHandler(
+                  false,
+                  new InlineCodeCompletionResponseEventListener(this));
+              requestHandler.call(new CallParameters(conversation,
```

Comment:
The `CompletionRequestHandler` wasn't really meant for use outside the scope of chat conversations. A better option would be to call the client directly, without all the SSE logic overhead.

For example:
```java
// CompletionRequestService.java

public @Nullable String getCodeCompletion(String prompt) {
  var selectedService = SettingsState.getInstance().getSelectedService();
  if (selectedService == OPENAI || selectedService == AZURE) {
    var request = CompletionRequestProvider.buildOpenAICodeCompletionRequest(prompt);
    var response = selectedService == OPENAI
        ? CompletionClientProvider.getOpenAIClient().getChatCompletion(request)
        : CompletionClientProvider.getAzureClient().getChatCompletion(request);
    return response.getChoices().get(0)
        .getMessage()
        .getContent();
  }
  if (selectedService == LLAMA_CPP) {
    return CompletionClientProvider.getLlamaClient()
        .getChatCompletion(CompletionRequestProvider.buildLlamaCodeCompletionRequest(prompt))
        .getContent();
  }

  return null;
}
```

Usage:
```java
var response = CompletionRequestService.getInstance().getCodeCompletion(prompt);
if (response != null) {
  ApplicationManager.getApplication().invokeLater(() -> createInlay(response));
  var insertAction = new InsertInlineTextAction(INLAY_KEY, response);
  ...
}
```

Reply:
Thanks for the suggestion, I thought I would keep it simple for the first draft to requesting completions, but you are right, there is no benefit/advantage for SSE here. 
I tried to implement it with synchronous calls as you suggested, but then the UI freezes so I guess I should use the `ProgressManager` to run the HTTP Request in Background Task:
```java
ProgressManager.getInstance().run(new Backgroundable(editor.getProject(), "CodeGPT Code Completion", true) {
  @Override
  public void run(@NotNull ProgressIndicator indicator) {
    var response = CompletionRequestService.getInstance()
        .getCodeCompletion(new CallParameters(conversation,
            ConversationType.INLINE_COMPLETION, message, false));
    if (response != null) {
      var inlineCode = response.replace(
          SettingsState.getInstance().getSelectedService().getFillInTheMiddle().getEot(),
          "");

      var insertAction = new InsertInlineTextAction(INLAY_KEY, inlineCode);
      ActionManager.getInstance().registerAction(ACTION_ID, insertAction);
      var keymap = KeymapManager.getInstance().getActiveKeymap();
      keymap.addShortcut(ACTION_ID,
          new KeyboardShortcut(KeyStroke.getKeyStroke(KeyEvent.VK_TAB, 0), null));
      ApplicationManager.getApplication().invokeLater(() -> createInlay(inlineCode));
    }
    restartTimer();
  }

  @Override
  public void onCancel() {
    super.onCancel();
  }
});
```
But the problem then is there is **no way to cancel the request,** should the user move the cursor, close the file or whatever. And from what I can see the `llm-client` only offers asynchronous calls with SSE right?
So maybe we should extend the `llm-client`s Clients to provide an async Request which return something like `CompletableFuture<OpenAIChatCompletionResponse`, that can also be cancelled?
I do not know what the best practice for implementing such async requests with `OkHttp`  is.

Reply:
I don't think there's a need for a progress manager indicator because in theory, the code completions need to be blazingly fast. However, it would be nice to have when local inference is slow.

Also, I would consider aborting the request to be a lower priority, and this PR can be merged without it. You can run long-running tasks on a pooled thread: `ApplicationManager.getApplication().executeOnPooledThread(() -> {});`

Additionally, you can use [MethodNameLookupListener.java](https://github.com/carlrobertoh/CodeGPT/blob/master/src/main/java/ee/carlrobert/codegpt/completions/MethodNameLookupListener.java#L41) as a reference; it already does a similar thing.

Reply:
Since we are now back to using SSE for code-completion, I added a `ProgressIndicator` again, with the option to cancel the completion request. Especially when running the LLM locally, its good for the user to see that a completion is going on and can cancel it.
I also added tests for `CodeCompletionService`.